src/External-contrib/NGFC/SOFC_ROM_P.py

```diff
@@ -0,0 +1,277 @@
+##############################################################################
+# The development of this flowsheet/code is funded by the ARPA-E DIFFERENTIATE
+# project: “Machine Learning for Natural Gas to Electric Power System Design”
+# Project number: DE-FOA-0002107-1625.
+# This project is a collaborative effort between the Pacific Northwest National
+# Laboratory, the National Energy Technology Laboratory, and the University of
+# Washington to design NGFC systems with high efficiencies and low CO2
+# emissions.
+##############################################################################
+##############################################################################
+# Institute for the Design of Advanced Energy Systems Process Systems
+# Engineering Framework (IDAES PSE Framework) Copyright (c) 2018-2020, by the
+# software owners: The Regents of the University of California, through
+# Lawrence Berkeley National Laboratory,  National Technology & Engineering
+# Solutions of Sandia, LLC, Carnegie Mellon University, West Virginia
+# University Research Corporation, et al. All rights reserved.
+#
+# Please see the files COPYRIGHT.txt and LICENSE.txt for full copyright and
+# license information, respectively. Both files are also available online
+# at the URL "https://github.com/IDAES/idaes-pse".
+##############################################################################
+"""
+This file builds the kriging SOFC reduced order model (developed by Pacific
+Northwest National Laboratory) in a Pyomo block. A data file of kriging
+coefficients must be provided. The ROM is specifically designed for use with
+the NGFC flowsheet. The code for generating SOFC ROMS can be found on PNNL's
+Github page: https://github.com/NGFC-Lib/NGFC-Lib.
+"""
+
+from pyomo.environ import Constraint, Param, Var, exp, value
+from pyomo.util.calc_var_value import calculate_variable_from_constraint
+
+
+# creates a dictionary from a list of indices and values
+def build_dict(index, values):
+    d = {}
+    for i, v in zip(index, values):
+        d[i] = v
+    return d
+
+
+# creates a dictionary with double indices
+def build_matrix(index1, index2, values):
+    d = {}
+    count = 0
+    for i in index1:
+        for j in index2:
+            d[(i, j)] = values[count]
+            count += 1
+    return d
+
+
+def make_SOFC_ROM(self):
+    # load kriging coefficients
+    ROM_filename = 'ADVNGFCNOCCS.dat'
```

Comment:
As @jghouse88 mentioned, as comment on the format of the data file might be useful, and perhaps a more meaningful name for the file (`ADVNGFCNOCCS.dat` does not really tell me what the file is for).